mmgen/models/gans/cyclegan.py

```diff
@@ -146,322 +61,169 @@ def get_module(self, module):
+            # GAN loss for generators
+            fake_pred = discriminators[domain](outputs[f'fake_{domain}'])
+            losses[f'loss_gan_g_{domain}'] = self.gan_loss(
+                fake_pred, target_is_real=True, is_disc=False)
+
+        # gen auxiliary loss
+        if self.with_gen_auxiliary_loss:
+            for loss_module in self.gen_auxiliary_losses:
+                loss_ = loss_module(outputs)
+                if loss_ is None:
+                    continue
+                # the `loss_name()` function return name as 'loss_xxx'
+                if loss_module.loss_name() in losses:
+                    losses[loss_module.loss_name(
+                    )] = losses[loss_module.loss_name()] + loss_
+                else:
+                    losses[loss_module.loss_name()] = loss_
 
         loss_g, log_vars_g = self._parse_losses(losses)
-        loss_g.backward()
 
-        return log_vars_g
+        return loss_g, log_vars_g
+
+    def _get_opposite_domain(self, domain):
+        for item in self._reachable_domains:
+            if item != domain:
+                return item
+        return None
 
-    def train_step(self, data_batch, optimizer):
+    def train_step(self,
+                   data_batch,
+                   optimizer,
+                   ddp_reducer=None,
+                   running_status=None):
         """Training step function.
 
         Args:
             data_batch (dict): Dict of the input data batch.
             optimizer (dict[torch.optim.Optimizer]): Dict of optimizers for
                 the generators and discriminators.
+            ddp_reducer (:obj:`Reducer` | None, optional): Reducer from ddp.
+                It is used to prepare for ``backward()`` in ddp. Defaults to
+                None.
+            running_status (dict | None, optional): Contains necessary basic
+                information for training, e.g., iteration number. Defaults to
+                None.
 
         Returns:
             dict: Dict of loss, information for logger, the number of samples\
                 and results for visualization.
         """
-        # data
-        img_a = data_batch['img_a']
-        img_b = data_batch['img_b']
-        meta = data_batch['meta']
+        # get running status
+        if running_status is not None:
+            curr_iter = running_status['iteration']
+        else:
+            # dirty walkround for not providing running status
+            if not hasattr(self, 'iteration'):
+                self.iteration = 0
+            curr_iter = self.iteration
 
         # forward generators
-        outputs = self.forward(img_a, img_b, meta, test_mode=False)
+        outputs = dict()
+        for target_domain in self._reachable_domains:
+            # fetch data by domain
+            source_domain = self.get_other_domains(target_domain)[0]
+            img = data_batch[f'img_{source_domain}']
+            # translation process
+            results = self.forward(
```

Comment:
Since this is an instance of `nn.Module`, you would better use `self(xx)` instead of `self.forward()`.